Add support for logging the trace-id in webapp2 apps.

csilvers opened the pull request:
Code was added in #3448 that set the trace-id when used in the
flask or django web frameworks.

This extends that to also support webapp2.

webapp2 is not cutting edge technology anymore -- it doesn't even work
on python3 -- but a lot of legacy GAE classic apps use webapp2, so I
think it makes sense to support it.

I tested this by running `nox` in the `logging/` directory, which as far
as I can tell is current state of the art for running tests.

csilvers commented:
Our company (Khan Academy) has signed the CLA.  I've been added to the group of authorized contributors.  Maybe I need to be added under my github.com email and not my company email.  I've done that and will try again.

csilvers commented:
(I made a new comment but it doesn't look like the clabot has run again.  I'm not sure what else I need to do.)

duggelz commented:
@lukesneeringer How do we poke the clabot?  (Craig is/was Google employee `#1` so I'll vouch for him if need be :)

duggelz commented:
The clabot doesn't think that Khan Academy has signed the corporate CLA (unless it has some wierd legal name that doesn't include the words Khan or Academy), nor does it think that csilvers (GitHub username) or csilvers@khanacademy.org has signed the individual CLA.

The docs says that corporate CLAs might take a few days to process, was it signed recently?

csilvers commented:
Ah, that sounds like it must be it.  I just signed the corporate CLA earlier today.  I'm happy to wait (though ideally it wouldn't need to block review!)

csilvers commented:
btw, hi doug!

duggelz commented:
Hi Craig! :)

dhermes commented on logging/google/cloud/logging/handlers/_helpers.py:
```diff
@@ -22,10 +22,20 @@
+    # If you try to import webapp2 under python3, you'll get a syntax
+    # error (since it hasn't been ported yet).  We just pretend it
+    # doesn't exist.  This is unlikely to hit in real life but does
+    # in the tests.
```
Ouch

dhermes commented on logging/google/cloud/logging/handlers/_helpers.py:
```diff
@@ -69,6 +79,33 @@ def get_trace_id_from_flask():
+    try:
+        # get_request() succeeds if we're in the middle of a webapp2
+        # request, or raises an assertion error otherwise:
+        # "Request global variable is not set".
```
`webapp2` has some real warts, eh?

csilvers replied:
Yeah, the AssertionError thing isn't part of the API :-), and I'm kinda worried that one day webapp2 will get updated in a way that raises a different exception.  But I think the tests will notice if that ever happens.

dhermes commented on logging/google/cloud/logging/handlers/_helpers.py:
```diff
@@ -69,6 +79,33 @@ def get_trace_id_from_flask():
+    """Get trace_id from webapp2 request headers.
+
+    :rtype: str
+    :return: Trace_id in HTTP request headers.
```
s/Trace_id/Trace ID?

csilvers replied:
OK, I was just copying from what the other functions in this file did.  I'll fix them all.  (Similarly for trace_id below.)

dhermes commented on logging/google/cloud/logging/handlers/_helpers.py:
```diff
@@ -95,7 +132,8 @@ def get_trace_id():
     """
-    checkers = (get_trace_id_from_django, get_trace_id_from_flask)
+    checkers = (get_trace_id_from_django, get_trace_id_from_flask,
+                get_trace_id_from_webapp2)
```
Nit: Can we just bite the bullet and make it:

```python
checkers = (
    get_trace_id_from_django,
    get_trace_id_from_flask,
    get_trace_id_from_webapp2,
)
```

csilvers replied:
Bullet bitten!  (Well, it will be.)

dhermes commented on logging/google/cloud/logging/handlers/_helpers.py:
```diff
@@ -69,6 +79,33 @@ def get_trace_id_from_flask():
+    if header is None:
+        return None
+
+    trace_id = header.split('/', 1)[0]
```
Should there be a helper for splitting the header in this way by now?

csilvers replied:
It's only one line of code, so I was happy copying it, but I can make a helper function too if you think it's clearer.

dhermes commented on logging/nox.py:
```diff
@@ -36,7 +36,7 @@ def unit_tests(session, python_version):
     session.install(
         'mock', 'pytest', 'pytest-cov',
-        'flask', 'django', *LOCAL_DEPS)
+        'flask', 'webapp2', 'webob', 'django', *LOCAL_DEPS)
```
Does installing these fail in Python 3? Worth conditionally adding the deps in Python 2 only?

csilvers replied:
They install fine in python3, you just can't use them.  I figure that's fine -- it lets us check that SyntaxError thing in the test -- so I'm happy enough to leave it like this.

But if you'd prefer to conditionally install the deps just for python2 -- under the theory nobody is going to be running python3 with webapp installed, so we could take out the SyntaxError handling -- that would make sense to me too.  I just don't know the syntax for it. :-)

dhermes commented on logging/tests/unit/handlers/test__helpers.py:
```diff
@@ -63,6 +64,57 @@ def test_valid_context_header(self):
 
+# webapp2 has not been ported to python3 yet, so do not try to test its
+# functionality there.
+if sys.version_info[0] == 2:
```
Instead, use `unittest.skipIf(six.PY3, 'webapp2 is Python 2 only')`

csilvers replied:
Nice!  I figured I wasn't doing it in the best way.  Though I do think that the interpreter still tries to byte-compile the test in this case?  Which means that in python3 it will still try to import webapp2 and break.  I guess if we do your other suggestion we'll catch the problem that way.

dhermes commented on logging/tests/unit/handlers/test__helpers.py:
```diff
@@ -63,6 +64,57 @@ def test_valid_context_header(self):
+    class Test_get_trace_id_from_webapp2(unittest.TestCase):
+
+        @staticmethod
+        def create_app():
```
Why not just make this a function at module scope?

csilvers replied:
I was just following what the existing tests here do.  I have no strong feelings myself.  I do feel it's nice to keep all the logic for a test encapsulated in the test class, to make it clear what helper objects go with what test, but I'm happy to change it if you have a preference.

dhermes commented on logging/tests/unit/handlers/test__helpers.py:
```diff
@@ -63,6 +64,57 @@ def test_valid_context_header(self):
+            return app
+
+        def setUp(self):
+            self.app = self.create_app()
```
Is there a **need** to do this in `setUp`? If not, (strongly) prefer just creating a test local `app` in each test case.

csilvers replied:
Again, I was just copying what the other tests did.  I will change them all to create a test-local `app`.

dhermes commented on logging/tests/unit/handlers/test__helpers.py:
```diff
@@ -63,6 +64,57 @@ def test_valid_context_header(self):
+        def create_app():
+            import webapp2
+
+            class GetTraceId(webapp2.RequestHandler):
```
Ditto here, define this handler at module scope?

```python
try:
    from webapp2 import RequestHandler
except (ImportError, SyntaxError):
    RequestHandler = object


class GetTraceId(RequestHandler):
    ...
```

csilvers replied:
That would work, but it seems a bit ugly to have to repeat the exception-list here.  That said, if I go to unittest.skip it's probably going to be necessary.

dhermes commented on logging/tests/unit/handlers/test__helpers.py:
```diff
@@ -63,6 +64,57 @@ def test_valid_context_header(self):
+                    from google.cloud.logging.handlers import _helpers
+
+                    trace_id = _helpers.get_trace_id_from_webapp2()
+                    self.response.content_type = "text/plain"
```
Nit: Single quotes instead of double quotes

csilvers replied:
:+1:

dhermes commented on logging/tests/unit/handlers/test__helpers.py:
```diff
@@ -63,6 +64,57 @@ def test_valid_context_header(self):
+            response = req.get_response(self.app)
+            trace_id = response.body
+
+            self.assertEquals("None", trace_id)
```
1. Why the string `None` versus the singleton?
1. Single quotes preferred

csilvers replied:
(1) is just a side-effect of how I wrote the test: I had the handler return the trace-id as text.  If the trace-id is None, then it gets auto-stringified to `None`.  I don't think webapp2 has something like flask's `test_request_context` so we have to go through a string intermediary at some point.

I could return json instead of just a raw text string, and then the json would convert `null` back to `None` -- would that be better?

csilvers replied:
(I'll do it that way.)

dhermes commented on logging/tests/unit/handlers/test__helpers.py:
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import sys
```
This import can go when you `import six` and use `six.PY3`

csilvers replied:
:+1:

csilvers commented on logging/google/cloud/logging/handlers/_helpers.py:
```diff
@@ -22,10 +22,20 @@
     _get_django_request)
 
 _FLASK_TRACE_HEADER = 'X_CLOUD_TRACE_CONTEXT'
+_WEBAPP2_TRACE_HEADER = 'x-cloud-trace-context'
```
Sure thing, will do.

liyanhui1228 commented:
@csilvers Yeah there is some inconsistent in the repo for `return` and `returns`, I confirmed with Jon that we should use `returns`. Sorry for the confusion.

Also I have just tested your changes in real life and it works well! I know it will take some time to figure out how to test it in real life because I did the same thing when I worked on this feature. Thank you again for sending this PR :)

csilvers commented:
Thanks @liyanhui1228 for doing the real-life testing!  And also for your code: I got compliments on my coding style and that's entirely because I cribbed from you. :-)

csilvers commented:
My latest commit addresses all the issues but these, which I'm leaving as-is unless people feel strongly:
* I didn't use a helper function for extracting a trace-id from the trace-id/span-id pair.
* I kept the code that installs webapp2 even for python3 tests (which ignore it).

liyanhui1228 commented:
This LGTM, but wait for @dhermes for merging. Thanks!

csilvers commented:
@liyanhui1228 do you mind approving it?  (However you do that in github.) It still has you marked as requesting changes.

lukesneeringer commented:
Good to merge. Ignore the CLA bot.

duggelz commented:
CLA bot now lists Khan Academy as a signer.
